Bug 1830080: Prompt user to update traffic on revision deletion

jeff-phillips-18 opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/ODC-2413

**Analysis / Root cause**: 
Users are allowed to delete revisions w/o adjusting traffic distribution.
Users are allowed to delete the last revision in a knative service.

**Solution Description**: 
When a revision is deleted, check if it is the only revision in the service, if so prevent the deletions and inform the user.

When a revision is ok to delete, prompt the user to update the traffic distribution prior to deleting the revision.

**Screen shots / Gifs for design review**:
Single Revision being deleted:
![image](https://user-images.githubusercontent.com/11633780/80754782-a3577c00-8afd-11ea-9a60-2fa4a485129d.png)

One of three revisions being deleted:
![image](https://user-images.githubusercontent.com/11633780/80754818-b1a59800-8afd-11ea-9483-6c61ead82a06.png)

One of two revisions being deleted where the second currently receives no traffic:
![image](https://user-images.githubusercontent.com/11633780/80754841-b9fdd300-8afd-11ea-8190-cba61adebcfc.png)

Process of deleting a revision:
![DeleteRev](https://user-images.githubusercontent.com/11633780/80755728-54aae180-8aff-11ea-99b0-97134343f1fb.gif)

**Browser conformance**: 
- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

cc @openshift/team-devconsole-ux @serenamarie125 @beaumorley 

/kind bug

jeff-phillips-18 commented:
@openshift/team-devconsole-ux @serenamarie125 @beaumorley PTAL

invincibleJai commented:
@jeff-phillips-18 Thanks , it looks great. I verified for below scenarios

- If only one revision is there 
It works as expected , user is shown with alert

- if there is more then one revision and user is trying to delete revision which has active traffic
It works as expected , user is shown with traffic distribution modal prior deletion

- if there is more then one revision and user is trying to delete revision which has **no active traffic**
Here as well user is shown with traffic distribution modal prior deletion, it makes sense as user is shown with pre-filled value of split

invincibleJai commented on frontend/packages/knative-plugin/src/components/revisions/DeleteRevisionModal.tsx:
```diff
@@ -0,0 +1,54 @@
+import { RevisionModel } from '../../models';
+import TrafficSplittingFields from '../traffic-splitting/TrafficSplittingFields';
+
+export interface TrafficSplittingDeleteModalProps {
```
nit: we may not need to export type

invincibleJai commented on frontend/packages/knative-plugin/src/components/revisions/DeleteRevisionModal.tsx:
```diff
@@ -0,0 +1,54 @@
+import TrafficSplittingFields from '../traffic-splitting/TrafficSplittingFields';
+
+export interface TrafficSplittingDeleteModalProps {
+  revisionItems: any;
```
we can type for `revisionItems`

jeff-phillips-18 commented:
Updated to not show the traffic splitting when the revision being deleted is not receiving traffic:

![image](https://user-images.githubusercontent.com/11633780/81186733-2acf4000-8f81-11ea-869f-0afeb7065375.png)

Updated error dialog when attempting to delete the last revision:

![image](https://user-images.githubusercontent.com/11633780/81186843-4cc8c280-8f81-11ea-87ae-dbf3502d7f17.png)

jeff-phillips-18 commented:
@openshift/team-devconsole-ux @beaumorley

jeff-phillips-18 commented:
/test backend

beaumorley commented:
Hi @jeff-phillips-18 this looks good. For the case were the user has to adjust the split for the two remaining revisions, is it an option to disable the delete button until they equal 100? I don't know if we already have a pattern here to show the error like you do so just asking.  

Also could we more clearly spell out what they need to do?  **(A) Make sure you update the traffic distribution among the remaining revisions to equal 100**. This might be too conversational but I would want the 100 part at least. So  **(A) Update the traffic distribution among the remaining revisions to equal 100** <--are there units here?

jeff-phillips-18 commented:
@beaumorley Those issues should likely be address in https://issues.redhat.com/browse/ODC-2479. There are more changes to the traffic splitting portion's styling and logic going on there.
Could you comment in that issue instead, raise a separate issue, or should that be fixed here, in your opinion?

serenamarie125 commented:
hey @beaumorley !  All good points, we definitely need to improve this dialog in the future - FYI we've talked about making it more consistent with the Kiali dialog, but these things would need to be tackled in future releases.

invincibleJai commented:
/lgtm

christianvogt commented:
/approve
